### PR TITLE
ENH: Add DistanceThreshold parameter in EuclideanDistance Metricv4

### PR DESCRIPTION
**Pull request by PranjalSahu**
Points beyond the DistanceThreshold parameter are skipped in the derivative and value calculation.
The code change has no impact on existing APIs as the default value is negative which will include all the points in the calculation. Only when a positive value is used as the DistanceThreshold we see a change in behavior. 

This change is to make the ICP registration similar to the Open3D API (**max_correspondence_distance**).

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

**Review by dzenanz — APPROVED** [review]

File: Modules/Registration/Metricsv4/test/itkEuclideanDistancePointSetMetricTest3.cxx [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+  using TranslationTransformType = itk::TranslationTransform<double, Dimension>;
+  auto translationTransform = TranslationTransformType::New();
+
+  // Instantiate the metric
+  using PointSetMetricType = itk::EuclideanDistancePointSetToPointSetMetricv4<PointSetType>;
+  auto metric = PointSetMetricType::New();
+  metric->SetFixedPointSet(fixedPoints);
+  metric->SetMovingPointSet(movingPoints);
+  metric->SetMovingTransform(translationTransform);
+  metric->SetDistanceThreshold(distanceThreshold);
```
Comment (jhlegarreta) [inline Modules/Registration/Metricsv4/test/itkEuclideanDistancePointSetMetricTest3.cxx:123]: Please exercise the `GetDistanceThreshold` method here using the `ITK_TEST_SET_GET_VALUE` macro.

File: Modules/Registration/Metricsv4/test/itkEuclideanDistancePointSetMetricTest3.cxx [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+  auto pointsLocator = PointsLocatorType::New();
+
+  auto fixedPoints = PointSetType::New();
+  fixedPoints->Initialize();
+
+  auto movingPoints = PointSetType::New();
+  movingPoints->Initialize();
+
+  // Threshold used to test the distance threshold functionality
+  double distanceThreshold = 0.5;
```
Comment (jhlegarreta) [inline Modules/Registration/Metricsv4/test/itkEuclideanDistancePointSetMetricTest3.cxx:48]: Are all the added `if/else` cases concerning the distance threshold value exercised in this test? Maybe the ones not covered here are covered in other tests through the default value. But for the sake of completeness, I would make sure all are tested here, maybe by making the value of the threshold (and the expected output values, if necessary) be a parameter of the method.

**Review by jhlegarreta — CHANGES_REQUESTED** [review]

File: Modules/Registration/Metricsv4/test/itkEuclideanDistancePointSetMetricTest3.cxx (hunk 2 above)
Comment (jhlegarreta) [reply Modules/Registration/Metricsv4/test/itkEuclideanDistancePointSetMetricTest3.cxx]: :100: Thanks @PranjalSahu.

File: Modules/Registration/Metricsv4/test/itkEuclideanDistancePointSetMetricTest3.cxx [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+int
+itkEuclideanDistancePointSetMetricTest3(int, char *[])
+{
+  int result = EXIT_SUCCESS;
+
+  // Test for positive distance threshold
+  if (itkEuclideanDistancePointSetMetricTest3Run<2>(0.5) == EXIT_FAILURE)
+  {
+    std::cerr << "Failed for Dimension 2." << std::endl;
```
Comment (jhlegarreta) [inline Modules/Registration/Metricsv4/test/itkEuclideanDistancePointSetMetricTest3.cxx:212]: The message could be more specific; now they do not allow knowing which case caused an error
```
constexpr unsigned int Dimension2D = 2:
double distanceThreshold = 0.5;
(...)
std::cerr << "Failed for Dimension " << Dimension2D << " and distance threshold " << distanceThreshold << std::endl;
```

**Review by jhlegarreta — COMMENTED** [review]
Some comments to improve the style of the test.

**Review by jhlegarreta — APPROVED** [review]
I would still use the declared dimension constants for printing in the messages, and squash both commits into one.

**Review by jhlegarreta — APPROVED** [review]
@PranjalSahu :100:. Thanks.

**Review by thewtex — APPROVED** [review]
:100: :1st_place_medal: well done @PranjalSahu !

**Pull request merged**